src/storage/leveldb/store.ts

```diff
@@ -0,0 +1,501 @@
+import LevelHistoryIndex from './history_index';
+import LevelUtxoIndex from './utxo_index';
+import LevelWalletIndex from './wallet_index';
+import LevelTokenIndex from './token_index';
+import walletApi from '../../api/wallet';
+import transaction from '../../utils/transaction';
+import fs from 'fs';
+import { HATHOR_TOKEN_CONFIG } from '../../constants';
+
+export default class LevelDBStore implements IStore {
+  addressIndex: LevelAddressIndex;
+  historyIndex: LevelHistoryIndex;
+  utxoIndex: LevelUtxoIndex;
+  walletIndex: LevelWalletIndex;
+  tokenIndex: LevelTokenIndex;
+  dbpath: string;
+  xpubkey: string;
+
+  constructor(dbroot: string, xpubkey: string) {
+    // The xpubkey in the account or change path?
+    this.xpubkey = xpubkey;
+    const xpub = HDPublicKey.fromString(xpubkey);
+    const dbpath = path.join(dbroot, xpub.publicKey.toString());
+    this.addressIndex = new LevelAddressIndex(dbpath);
+    this.historyIndex = new LevelHistoryIndex(dbpath);
+    this.utxoIndex = new LevelUtxoIndex(dbpath);
+    this.walletIndex = new LevelWalletIndex(dbpath);
+    this.tokenIndex = new LevelTokenIndex(dbpath);
+
+    this.dbpath = dbpath;
+  }
+
+  async close(): Promise<void> {
+    await this.addressIndex.close();
+    await this.historyIndex.close();
+    await this.utxoIndex.close();
+    await this.walletIndex.close();
+    await this.tokenIndex.close();
+  }
+
+  async destroy(): Promise<void> {
+    await this.close();
+    if (process.version < 'v14.14.0') {
+      // rmSync was introduced in v14.14.0, earlier versions should use rmdirSync
+      // https://nodejs.org/docs/latest-v14.x/api/fs.html#fs_fs_rmdirsync_path_options
+      fs.rmdirSync(this.dbpath, { recursive: true });
+    } else {
+      fs.rmSync(this.dbpath, { recursive: true, force: true });
+    }
+  }
+
+  async validate(): Promise<void> {
+    await this.addressIndex.validate();
+    await this.historyIndex.validate();
+    await this.utxoIndex.validate();
+    await this.tokenIndex.validate();
+    await this.walletIndex.validate();
+  }
+
+  async *addressIter(): AsyncGenerator<IAddressInfo> {
+    for await (const info of this.addressIndex.addressIter()) {
+      yield info;
+    }
+  }
+
+  async getAddress(base58: string): Promise<IAddressInfo | null> {
+    return this.addressIndex.getAddressInfo(base58);
+  }
+
+  async getAddressMeta(base58: string): Promise<IAddressMetadata | null> {
+    return this.addressIndex.getAddressMeta(base58);
+  }
+
+  async addressCount(): Promise<number> {
+    return this.addressIndex.addressCount();
+  }
+
+  async getAddressAtIndex(index: number): Promise<IAddressInfo | null> {
+    const address = await this.addressIndex.getAddressAtIndex(index);
+    if (address === null) {
+      return null;
+    }
+    return this.addressIndex.getAddressInfo(address);
+  }
+
+  async saveAddress(info: IAddressInfo): Promise<void> {
+    if (!info.base58) {
+      throw new Error('Invalid address');
```

Comment:
Should we have standard errors like the ones in errors.ts so they can be better handled?

e.g.

`class InvalidAddressError extends Error`

Reply:
I also believe we should create a standard error, but since it would affect both the persistent storage files and the base storage scheme implementation I think we should make this on its own PR, what do you think?

Reply:
I agree, can you create an issue so we don't forget to refactor this?